lib/AST/AlignTokens.cpp

```diff
@@ -2039,368 +1799,213 @@ Result<std::monostate, std::string> ASTImpl::AlignTokens(
-          work_list.push_back(sub_decl);
-        }
+    // Find the end of the hole.
+    auto end_of_hole = &(tok[1]);
+    for (; end_of_hole < printed_end; ++end_of_hole) {
+      if (end_of_hole->opaque_source_loc != TokenImpl::kInvalidSourceLocation) {
         break;
+      }
+    }
 
-//      // If it's something like `extern "C" int foo;` then we want to treat it
-//      // as top-level, otherwise, it's more like `extern "C" { ... }` and so we
-//      // want to find the top-level decls in the `...`.
-//      case clang::Decl::LinkageSpec:
-//        if (auto lsp = llvm::dyn_cast<clang::LinkageSpecDecl>(decl);
-//            !lsp->hasBraces() && !lsp->isImplicit()) {
-//
-//          // Compute bounds of top-level decls. This will fill out
-//          // `ast->lexically_containing_decl`.
-//          (void) ast->DeclBounds(decl);
-//          tlds.push_back(decl);
-//
-//        } else {
-//          for (auto sub_decl : clang::Decl::castToDeclContext(decl)->decls()) {
-//            work_list.push_back(sub_decl);
-//          }
-//        }
-//        break;
+    // There is no hole.
+    if (end_of_hole == &(tok[1])) {
+      continue;
+    }
 
-      default:
-        if (auto ftpl = clang::dyn_cast<clang::FunctionTemplateDecl>(decl)) {
-          for (clang::Decl *spec : ftpl->specializations()) {
-            ignore_decls.insert(Canonicalize(spec));
-          }
+    // Convert our printed token hole bounds into parsed token hole bounds.
+    auto parsed_it = loc_to_toks.find(tok->opaque_source_loc);
+    auto parsed_end_it = loc_to_toks.find(end_of_hole->opaque_source_loc);
+    if (parsed_it == loc_to_toks.end() || parsed_end_it == loc_to_toks.end()) {
+      continue;
+    }
 
-        } else if (auto ctpl = clang::dyn_cast<clang::ClassTemplateDecl>(decl)) {
-          for (clang::Decl *spec : ctpl->specializations()) {
-            ignore_decls.insert(Canonicalize(spec));
-          }
+    auto first_parsed_tok = parsed_it->second;
+    auto last_parsed_tok = parsed_end_it->second;
 
-        } else if (auto vtpl = clang::dyn_cast<clang::VarTemplateDecl>(decl)) {
-          for (clang::Decl *spec : vtpl->specializations()) {
-            ignore_decls.insert(Canonicalize(spec));
-          }
-        }
+    // Make sure our hole sized match.
+    if (first_parsed_tok >= last_parsed_tok ||
+        (last_parsed_tok - first_parsed_tok) != (end_of_hole - tok)) {
+      continue;
+    }
 
-        if (!decl->isImplicit()) {
-          // Compute bounds of top-level decls. This will fill out
-          // `ast->lexically_containing_decl`.
-          (void) ast->DeclBounds(decl);
-          tlds.push_back(decl);
-        }
-        break;
+    // Force merge on the hole.
+    bool changed = false;
+    for (auto hole_tok = tok; hole_tok < end_of_hole; ) {
+      ++first_parsed_tok;
+      ++hole_tok;
+      MergeToken(first_parsed_tok, hole_tok, changed, true  /* force */);
     }
+
+    assert(changed);
+    (void) changed;
```

Comment:
This. Also, why is it even here as the `changed` is used?

Reply:
I will admit that I no longer fully understand when `changed` matters and when it doesn't. Figuring out a better fixpoint progress tracking mechanism goes hand in hand with better APIs in here, and I think is beyond the scope of this PR.